sparkle/output/selection_output.py

```diff
@@ -0,0 +1,229 @@
+        """Initalize SelectionSolverData.
+
+        Args:
+            solver_performance_ranking: list with solvers ranked by avg. performance
+            num_solvers: The number of solvers
+        """
+        self.solver_performance_ranking = solver_performance_ranking
+        self.single_best_solver = solver_performance_ranking[0][0]
+        self.num_solvers = num_solvers
+
+
+class SelectionPerformance:
+    """Class that stores selection performance results."""
+    def __init__(self: SelectionSolverData,
+                 performance_path: Path,
+                 vbs_par: float, metric: str) -> None:
+        """Initalize SelectionPerformance.
+
+        Args:
+            performance_path: Path to portfolio selector performance
+            vbs_par: The performance of the virtual best selector
+            metric: The metric of the SparkleObjective
+        """
+        if not performance_path.exists():
+            print(f"ERROR: {performance_path} does not exist.")
+            sys.exit(-1)
+        actual_performance_data = PerformanceDataFrame(performance_path)
+        self.vbs_par = vbs_par
+        self.actual_PAR = actual_performance_data.mean()
+        self.metric = metric
+
+
+class SelectionOutput:
+    """Class that collects configuration data and outputs it a JSON format."""
+
+    def __init__(self: SelectionOutput, selection_scenario: Path,
+                 performance_data_path: Path, feature_path: Path,
+                 training_instances: list[InstanceSet],
+                 test_instances: list[InstanceSet],
+                 cutoff_time: int, penalised_time: int,
+                 output: Path = None) -> None:
+        """Initialize SelectionOutput class.
+
+        Args:
+            selection_scenario: Path to selection output directory
+            performance_data_path: Path to the selector performance data
+            feature_path: Path to the csv feature file
+            training_instances: The set of training instances
+            test_instances: The set of test instances
+            cutoff_time: The cutoff time
+            penalised_time: The penalised time
+            output: Path to the output directory
+        """
+        if output.is_dir():
+            self.output = output / "selection.json"
+        else:
+            self.output = output
+
+        self.training_instances = training_instances
+        self.test_instances = test_instances
+        self.cutoff_time = cutoff_time
+        self.penalised_time = penalised_time
+
+        train_data = PerformanceDataFrame(performance_data_path)
+        train_data.penalise(cutoff_time,
+                            penalised_time)
+        self.objective = SparkleObjective(train_data.objective_names[0])
+        self.solver_data = self.get_solver_data(train_data)
+
+        # Collect marginal contribution data
+        feature_data = FeatureDataFrame(feature_path)
+        self.rank_list_perfect = train_data.marginal_contribution(sort=True)
```

Comment:
rank list is just for report generation, here you can just call it self.marginal_contribution_perfect and self.marginal_contribution_actual